clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -47,32 +190,166 @@ public boolean isValid() {
         }
     }
 
-    public TransactionState(Time time) {
-        this.pidAndEpoch = new PidAndEpoch(NO_PRODUCER_ID, NO_PRODUCER_EPOCH);
-        this.sequenceNumbers = new HashMap<>();
-        this.time = time;
+    public boolean hasPendingTransactionalRequests() {
```

Comment:
Both of these should only be accessed from the background sender thread, so there should be no contention.